cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/events/KubernetesRestartEventPublisher.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.resource.events;
+
+import io.fabric8.kubernetes.api.model.MicroTime;
+import io.fabric8.kubernetes.api.model.ObjectReference;
+import io.fabric8.kubernetes.api.model.ObjectReferenceBuilder;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.strimzi.operator.cluster.model.RestartReason;
+import io.strimzi.operator.cluster.model.RestartReasons;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.time.Clock;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Locale;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+/**
+ * Publishes KafkaRoller restart events as Kubernetes events, using either the events.k8s.io/v1beta1 or v1 API,
+ * depending on which is available.
+ */
+public abstract class KubernetesRestartEventPublisher {
+
+    private static final Logger LOG = LogManager.getLogger(KubernetesRestartEventPublisher.class);
+
+    private final Clock clock;
+
+    private static final DateTimeFormatter K8S_MICROTIME = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'.'SSSSSSXXX");
+
+    protected String action = "StrimziInitiatedPodRestart";
+    protected String controller = "strimzi.io/cluster-operator";
+
+    // Matches first character or characters following an underscore
+    private static final Pattern PASCAL_CASE_HELPER = Pattern.compile("^.|_.");
+
+    // K8s events are required to have a message of 1KiB or smaller
+    private static final int MAX_MESSAGE_LENGTH = 1000;
+    private static final String ELLIPSIS = "...";
+
+    public KubernetesRestartEventPublisher() {
+        this(Clock.systemDefaultZone());
+    }
+
+    protected KubernetesRestartEventPublisher(Clock clock) {
+        this.clock = clock;
+    }
+
+    /**
+     * Create an instance for the highest event API, there's three possible subtypes, due to how fabric8 delineates the differing
+     * event APIs.
+     *
+     * @param client       Kubernetes client
+     * @param operatorName the pod name of the current cluster operator instance
+     * @param hasEventsV1  if the cluster is using events.k8s.io/v1 instead of events.k8s.io/v1beta1
+     * @return instance of the appropriate publisher for the given API
+     */
+    public static KubernetesRestartEventPublisher createPublisher(KubernetesClient client, String operatorName, boolean hasEventsV1) {
+        Clock clock = Clock.systemDefaultZone();
+        if (hasEventsV1) {
+            return new V1RestartEventPublisher(clock, client, operatorName);
```

Comment:
Would it be better to use this to inject an eventBuilder function and a `Consumer<Event>` to do the actual publication into `KubernetesRestartEventPublisher` rather than sub classing? 

It looks to me like the eventBuilder is common between the two sub classes so may just need the consumer.

Reply:
I used subclasses mainly to avoid Java's lack of type aliasing, as both event apis have the same type names, one of those functions would end up with package names throughout, decreasing readability.